Sources/NIOHTTP2/HTTP2StreamChannel.swift

```diff
@@ -338,7 +361,28 @@ final class HTTP2StreamChannel: Channel, ChannelCore {
         }
 
         let frame = self.unwrapData(data, as: HTTP2Frame.self)
+
+        // We need a promise to attach our flow control callback to.
+        // Regardless of whether the write succeeded or failed, we don't count
+        // the bytes any longer.
+        let promise = promise ?? self.eventLoop.makePromise()
+        let writeSize = frame.bufferBytes
+
+        // Right now we deal with this math by just attaching a callback to all promises. This is going
+        // to be annoyingly expensive, but for now it's the most straightforward approach.
+        promise.futureResult.hop(to: self.eventLoop).whenComplete { (_: Result<Void, Error>) in
+            if case .changed(newValue: let value) = self.writabilityManager.wroteBytes(writeSize) {
+                self._isWritable.store(value)
+                self.pipeline.fireChannelWritabilityChanged()
+            }
+        }
         self.pendingWrites.append((frame, promise))
+
+        // Ok, we can make an outcall now, which means we can safely deal with the flow control.
+        if case .changed(newValue: let value) = self.writabilityManager.bufferedBytes(writeSize) {
```

Comment:
nit: it seems like this code block and the one in the callback of the future is the same, so just make it a private function (can even be just declared in this function itself.